Purchases/Public/RCPurchases.m

```diff
@@ -734,50 +748,61 @@ - (void)offeringsWithCompletionBlock:(RCReceiveOfferingsBlock)completion
-                                                SKProduct *product = productsById[productIdentifier];
-
-                                                if (product == nil) {
-                                                    [missingProducts addObject:productIdentifier];
-                                                }
-                                            }];
-
-                                            if (missingProducts.count > 0) {
-                                                RCLog(@"Could not find SKProduct for %@", missingProducts);
-                                                RCLog(@"Ensure your products are correctly configured in App Store Connect");
-                                                RCLog(@"See https://www.revenuecat.com/2018/10/11/configuring-in-app-products-is-hard");
-                                            }
-                                            [self.deviceCache cacheOfferings:offerings];
-
-                                            CALL_AND_DISPATCH_IF_SET(completion, offerings, nil);
-                                        } else {
-                                            CALL_AND_DISPATCH_IF_SET(completion, nil, [RCPurchasesErrorUtils unexpectedBackendResponseError]);
-                                        }
+        NSMutableDictionary *productsById = [NSMutableDictionary new];
+        for (SKProduct *p in products) {
+            productsById[p.productIdentifier] = p;
+        }
+        RCOfferings *offerings = [self.offeringsFactory createOfferingsWithProducts:productsById data:data];
+        if (offerings) {
+            NSMutableArray *missingProducts = [NSMutableArray new];
+            [self performOnEachProductIdentifierInOfferings:data block:^(NSString *productIdentifier) {
+                SKProduct *product = productsById[productIdentifier];
+
+                if (product == nil) {
+                    [missingProducts addObject:productIdentifier];
+                }
+            }];
 
-                                    }];
-                                   }];
+            if (missingProducts.count > 0) {
+                RCLog(@"Could not find SKProduct for %@", missingProducts);
+                RCLog(@"Ensure your products are correctly configured in App Store Connect");
+                RCLog(@"See https://www.revenuecat.com/2018/10/11/configuring-in-app-products-is-hard");
+            }
+            [self.deviceCache cacheOfferings:offerings];
+
+            CALL_AND_DISPATCH_IF_SET(completion, offerings, nil);
+        } else {
+            CALL_AND_DISPATCH_IF_SET(completion, nil, [RCPurchasesErrorUtils unexpectedBackendResponseError]);
+        }
+
+    }];
+}
+
+- (void)handleOfferingsUpdateError:(NSError *)error completion:(RCReceiveOfferingsBlock)completion {
+    RCLog(@"Error fetching offerings - %@", error);
+    CALL_AND_DISPATCH_IF_SET(completion, nil, error);
+    [self.deviceCache clearOfferingsCacheTimestamp];
```

Comment:
Should we clear before doing `CALL_AND_DISPATCH_IF_SET`?

Reply:
great catch